src/__tests__/unit/controllers/hooks/router-store/test.tsx

```diff
@@ -0,0 +1,306 @@
+    await nextTick();
+
+    expect(fooVal).toEqual('hello');
+    expect(barVal).toEqual('world');
+    expect(wrapper.find(ComponentFoo).html()).toEqual(
+      '<div id="foo-count">2</div>'
+    );
+    expect(wrapper.find(ComponentBar).html()).toEqual(
+      '<div id="bar-count">2</div>'
+    );
+
+    act(() => fooUpdateFn('newVal'));
+    await nextTick();
+
+    // URL is now — /projects/123/board/456?foo=newVal&bar=world
+
+    expect(fooVal).toEqual('newVal');
+    expect(barVal).toEqual('world');
+    expect(wrapper.find(ComponentFoo).html()).toEqual(
+      '<div id="foo-count">3</div>'
+    );
+    expect(wrapper.find(ComponentBar).html()).toEqual(
+      '<div id="bar-count">2</div>'
+    );
+
+    act(() => barUpdateFn('newVal'));
+    await nextTick();
+
+    // URL is now — /projects/123/board/456?foo=newVal&bar=newVal
+
+    expect(fooVal).toEqual('newVal');
+    expect(barVal).toEqual('newVal');
+    expect(wrapper.find(ComponentFoo).html()).toEqual(
+      '<div id="foo-count">3</div>'
+    );
+    expect(wrapper.find(ComponentBar).html()).toEqual(
+      '<div id="bar-count">3</div>'
+    );
+  });
+});
+
+describe('usePathParam', () => {
+  beforeEach(() => {
+    history = historyHelper.createMemoryHistory(historyBuildOptions);
+    historyPushSpy = jest.spyOn(history, 'push');
+  });
+
+  afterEach(() => {
+    defaultRegistry.stores.clear();
+    jest.resetAllMocks();
+    jest.restoreAllMocks();
+  });
+
+  it('should return the right param value', () => {
+    let ppVal: string | undefined;
+
+    mount(
+      <Router routes={mockRoutes} history={history}>
+        <MockComponent>
+          {() => {
+            const [param] = usePathParam('projectId');
+            ppVal = param;
+
+            return <div>I am a subscriber</div>;
+          }}
+        </MockComponent>
+      </Router>
+    );
+    expect(ppVal).toEqual('123');
+  });
+
+  it('should return undefined for non-existent params', () => {
+    let ppVal: string | undefined;
+    mount(
+      <Router routes={mockRoutes} history={history}>
+        <MockComponent>
+          {() => {
+            const [param] = usePathParam('iamnotapathparam');
+            ppVal = param;
+
+            return <div>I am a subscriber</div>;
+          }}
+        </MockComponent>
+      </Router>
+    );
+    expect(ppVal).toEqual(undefined);
+  });
+
+  it('should update URL with new param value', async () => {
+    let qpVal: string | undefined, qpUpdateFn: (qp: string) => void;
+
+    mount(
+      <Router routes={mockRoutes} history={history}>
+        <MockComponent>
+          {() => {
+            const [param, setParam] = usePathParam('projectId');
+            qpVal = param;
+            qpUpdateFn = setParam;
+
+            return <div>I am a subscriber</div>;
+          }}
+        </MockComponent>
+      </Router>
+    );
+
+    expect(qpVal).toEqual('123');
+
+    act(() => qpUpdateFn('newVal'));
```

Comment:
what happens the update function is called with `undefined` or `null`? Can we add a test for that?

Reply:
We can't update the path to have null or undefined. Path compilation package we use throws an error in that case. And I think that's expected. Do you think we should filter out `null` or `undefined` path params before updating the path? @jackbrown

Reply:
We're working on making `usePathParam` to accept `undefined` in the setter. @albertogasparin pointed out that we can have optional path params and we're accommodating that case to our API.

Reply:
@albertogasparin @jackbrown `usePathParam` setter now takes `undefined` that will remove optional params from the url and throw when undefined is passed for non-optional params. Added tests around these cases as well.